Runtime/JSPromise.cs

```diff
@@ -0,0 +1,165 @@
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+using static NodeApi.JSNativeApi.Interop;
+
+namespace NodeApi;
+
+public readonly struct JSPromise : IEquatable<JSValue>
+{
+    private readonly JSValue _value;
+
+    public static explicit operator JSPromise(JSValue value) => new(value);
+    public static implicit operator JSValue(JSPromise promise) => promise._value;
+
+    public static explicit operator JSPromise(JSObject obj) => (JSPromise)(JSValue)obj;
+    public static implicit operator JSObject(JSPromise promise) => (JSObject)promise._value;
+
+    private JSPromise(JSValue value)
+    {
+        _value = value;
+    }
+
+    public delegate void Resolve(Action<JSValue> resolve);
+
+    public delegate Task AsyncResolve(Action<JSValue> resolve);
+
+    public delegate void ResolveReject(
+        Action<JSValue> resolve,
+        Action<JSValue> reject); // TODO: Change reject type argument to JSError?
+
+    public delegate Task AsyncResolveReject(
+        Action<JSValue> resolve,
+        Action<JSValue> reject); // TODO: Change reject type argument to JSError?
+
+    /// <summary>
+    /// Creates a new JS Promise with a resolve callback.
+    /// </summary>
+    /// <param name="callback">Callback that is invoked immediately and must _eventually_ invoke
+    /// either the resolve function with a <see cref="JSValue"/> or throw an exception.</param>
+    /// <remarks>
+    /// Any exception thrown by the callback will be caught and used as a promise rejection error.
+    /// </remarks>
+    public JSPromise(Resolve callback)
+    {
+        _value = JSValue.CreatePromise(out Deferred deferred);
+        try
+        {
+            callback(deferred.Resolve);
+        }
+        catch (Exception ex)
+        {
+            deferred.Reject(ex);
+        }
+    }
+
+    /// <summary>
+    /// Creates a new JS Promise with an async resolve callback.
+    /// </summary>
+    /// <param name="callback">Callback that is invoked immediately and must _eventually_ invoke
+    /// either the resolve function with a <see cref="JSValue"/> or throw an exception.</param>
+    /// <remarks>
+    /// Any (sync or async) exception thrown by the callback will be caught and used as a promise
+    /// rejection error.
+    /// </remarks>
+    public JSPromise(AsyncResolve callback)
+    {
+        _value = JSValue.CreatePromise(out Deferred deferred);
+        async void AsyncCallback()
+        {
+            using var __asyncScope = new JSAsyncScope();
```

Comment:
>__asyncScope [](http://example.com/codeflow?start=22&length=12)

Can we avoid using double underscore for the local variable names?

Reply:
Oh, that was a copy-paste mistake here. The double-underscore should only be used in generated code.